jujugui/static/gui/src/app/components/inspector-expose/inspector-expose.js

```diff
@@ -0,0 +1,138 @@
+
+YUI.add('inspector-expose', function() {
+
+  juju.components.InspectorExpose = React.createClass({
+
+    /**
+      The callable to be passed to the unit items for navigating to the unit
+      details.
+
+      @method _unitItemAction
+      @param {Object} e The click event.
+    */
+    _unitItemAction: function(e) {
+      var unitId = e.currentTarget.getAttribute('data-id').split('/')[1];
+      this.props.changeState({
+        sectionA: {
+          component: 'inspector',
+          metadata: {
+            id: this.props.service.get('id'),
+            unit: unitId,
+            activeComponent: 'unit'
+          }
+        }
+      });
+    },
+
+    /**
+      Generate a list of units for the service
+
+      @method _generateUnits
+      @returns {Array} A list of units.
+    */
+    _generateUnits: function() {
+      var units = [];
+      this.props.units.toArray().forEach(function(unit) {
+        units.push(
+          <juju.components.InspectorExposeUnit
+            key={unit.id}
+            action={this._unitItemAction}
+            unit={unit} />);
+      }, this);
+      return units;
+    },
+
+    /**
+      Display a list of units if the service is exposed.
+
+      @method _displayUnitList
+      @returns {Object} A list of units.
+    */
+    _displayUnitList: function() {
+      if (!this.props.service.get('exposed')) {
+        return;
+      }
+      return <ul className="inspector-expose__units">
+          {this._generateUnits()}
+        </ul>;
+    },
+
+    /**
+      Expose the service when toggled.
+
+      @method _handleExposeChange
+    */
+    _handleExposeChange: function() {
+      var service = this.props.service;
+      var serviceId = service.get('id');
+      if (service.get('exposed')) {
+        this.props.unexposeService(serviceId,
+            this._exposeServiceCallback, {});
+      } else {
+        this.props.exposeService(serviceId,
+            this._exposeServiceCallback, {});
+      }
+    },
+
+    /**
+      Callback to handle errors when exposing a service.
+
+      @method _exposeServiceCallback
+      @param {object} e The expose event
+    */
+    _exposeServiceCallback: function(e) {
+      // XXX: Need to implement expose error handling.
+      if (e.err) {
+        console.error(e.err);
+      }
+    },
+
+    render: function() {
+      var toggle = {
+        key: 'expose-toggle'
+      };
+      return (
+        <div className="inspector-expose">
+            <div className="inspector-expose__control">
+              <juju.components.BooleanConfig
+                key={toggle.key}
+                ref={toggle.key}
+                option={toggle}
+                onChange={this._handleExposeChange}
+                label="Expose service"
+                config={this.props.service.get('exposed')} />
+            </div>
+            <p className="inspector-expose__warning">
+              Exposing this service will make it publicly accessible from
+              the web
```

Comment:
I'm not sure this message is technically valid as all it does is open the ports.